File: packages/android/editor-schema/src/main/kotlin/app/notes/editor/Schema.kt
package app.notes.editor

/**
 * Mirror of the TS `@notes-app/editor-schema`. The two are kept in sync by
 * hand — there is no codegen because the schema is intentionally tiny and
 * any drift would be visible in the round-trip tests on both sides.
 */

const val SCHEMA_VERSION: Int = 1

enum class BlockNodeName(val wireName: String) {
    PARAGRAPH("paragraph"),
    HEADING("heading"),
    BULLET_LIST("bullet_list"),
    ORDERED_LIST("ordered_list"),
    CHECK_LIST("check_list"),
    CODE_BLOCK("code_block"),
    IMAGE("image"),
    ATTACHMENT("attachment"),
    ;

    companion object {
        fun fromWire(value: String): BlockNodeName? = entries.firstOrNull { it.wireName == value }
    }
}

enum class InlineNodeName(val wireName: String) {
    TEXT("text"),
    CODE_INLINE("code_inline"),
    INTERNAL_LINK("internal_link"),
    HARD_BREAK("hard_break"),
    ;

    companion object {
        fun fromWire(value: String): InlineNodeName? = entries.firstOrNull { it.wireName == value }
    }
}

enum class MarkName(val wireName: String) {
    BOLD("bold"),
    ITALIC("italic"),
    STRIKE("strike"),
    ;

    companion object {
        fun fromWire(value: String): MarkName? = entries.firstOrNull { it.wireName == value }
    }
}

/**
 * Lightweight document model. Mirrors the TS `DocNode` shape. We don't
 * pull in a JSON dependency at this layer — the Android shell (M12) wires
 * Moshi/Gson to (de)serialise this onto `notes.body_json`.
 */
data class DocNode(
    val type: String,
    val attrs: Map<String, Any?> = emptyMap(),
    val text: String? = null,
    val marks: List<Mark> = emptyList(),
    val content: List<DocNode> = emptyList(),
)

data class Mark(val type: MarkName)

/** A doc is just a DocNode of type "doc" — the helper makes intent explicit. */
data class Doc(val content: List<DocNode>) {
    fun toNode(): DocNode = DocNode(type = "doc", content = content)

    companion object {
        fun empty(): Doc = Doc(content = listOf(DocNode(type = "paragraph")))

        fun fromNode(node: DocNode): Doc {
            require(node.type == "doc") { "Doc.fromNode expected type=doc, got ${node.type}" }
            return Doc(content = node.content)
        }
    }
}

File: packages/editor-schema/src/schema.ts
/**
 * Frozen ProseMirror schema for notes-app. The schema is intentionally tiny
 * and conservative — adding a node is a deliberate decision that requires
 * a migration plan because every existing note must still render under the
 * new schema.
 *
 * The schema is declared as plain data here so both clients (Tiptap on the
 * web, the Compose renderer on Android) can build their concrete schema
 * instances from the same source. The Kotlin mirror lives at
 * `packages/android/editor-schema/.../Schema.kt`.
 */

/** Bump only when an incompatible node/mark is added or removed. */
export const SCHEMA_VERSION = 1;

/** The set of block-level node names allowed at the top of the document. */
export const BLOCK_NODE_NAMES = [
  "paragraph",
  "heading",
  "bullet_list",
  "ordered_list",
  "check_list",
  "code_block",
  "image",
  "attachment",
] as const;
export type BlockNodeName = (typeof BLOCK_NODE_NAMES)[number];

/** Inline node names that can appear inside a paragraph or heading. */
export const INLINE_NODE_NAMES = ["text", "code_inline", "internal_link", "hard_break"] as const;
export type InlineNodeName = (typeof INLINE_NODE_NAMES)[number];

export type NodeName = "doc" | "list_item" | "check_list_item" | BlockNodeName | InlineNodeName;

/** Marks applied to text runs. */
export const MARK_NAMES = ["bold", "italic", "strike"] as const;
export type MarkName = (typeof MARK_NAMES)[number];

export interface MarkInstance {
  readonly type: MarkName;
}

/** Shape every document node must satisfy. */
export interface DocNode {
  readonly type: NodeName;
  readonly attrs?: Readonly<Record<string, unknown>>;
  readonly content?: readonly DocNode[];
  readonly text?: string;
  readonly marks?: readonly MarkInstance[];
}

/** Top-level document. */
export interface DocRoot extends DocNode {
  readonly type: "doc";
  readonly content: readonly DocNode[];
}

/** Build an empty doc (the schema's identity element). */
export function emptyDoc(): DocRoot {
  return { type: "doc", content: [{ type: "paragraph" }] };
}

/** Cheap structural validation — full validation is in `validateDoc()`. */
export function isDocRoot(value: unknown): value is DocRoot {
  return (
    typeof value === "object" &&
    value !== null &&
    (value as { type?: unknown }).type === "doc" &&
    Array.isArray((value as { content?: unknown }).content)
  );
}
